Rent

joredudiaz opened the pull request:
# Summary
- model rent created
# Trello card
https://trello.com/c/zPR5b7eI

joredudiaz committed:
CREATE
app/models/rent.rb
rent model

db/migrate/20181026173627_create_rents.rb
rent migration

spec/models/rent_spec.rb
rent test

joredudiaz committed:
UPDATE
app/models/book.rb|user.rb
association with other models

app/models/rent.rb
-association with other models
-validations of fields

db/migrate/20181026173627_create_rents.rb
add references to the table

spec/models/rent_spec.rb
test field of the table

CREATE
spec/factories/rent_factory.rb
data for test

joredudiaz committed:
UPDATE test to save rent with valid and invalid params

joredudiaz committed:
UPDATE rubocop fix

holywyvern commented on app/models/rent.rb:
```diff
@@ -0,0 +1,15 @@
+
+  private
+
+  def end_date_greather_than_start_date
```
You may move this into ApplicationRecord or a concern.

holywyvern commented on db/migrate/20181026173627_create_rents.rb:
```diff
@@ -0,0 +1,13 @@
+    end
+
+    add_reference :rents, :user, foreign_key: true, null: false
+    add_reference :rents, :book, foreign_key: true, null: false
```
If you are adding references on the rents table, you may then add them on the same "create_table" block using `t.references`

holywyvern commented on spec/models/rent_spec.rb:
```diff
@@ -0,0 +1,24 @@
+      end
+    end
+
+    let(:emptyRent) { Rent.new }
```
You may check properties one at a time, like `build( :rent, book: null)` and check for that instead.

holywyvern commented on spec/models/rent_spec.rb:
```diff
@@ -0,0 +1,24 @@
+    let(:emptyRent) { Rent.new }
+    context 'With invalid params' do
+      it 'returns exception' do
+        expect { emptyRent.save! }.to raise_error(ActiveRecord::RecordInvalid)
```
You must use `expect(emptyRent).to be_invalid`

joredudiaz committed:
UPDATE
app/models/rent.rb
include validation created

app/validators/start_end_date_validator.rb
moved a validator into its own class

db/migrate/20181026173627_create_rents.rb
include reference on the definition of the table

spec/models/rent_spec.rb
- include validation of each field
- change expect to "be invalid"

joredudiaz committed:
UPDATE rubocop fix

leoncalermo commented on db/migrate/20181026173627_create_rents.rb:
```diff
@@ -0,0 +1,19 @@
+class CreateRents < ActiveRecord::Migration[5.2]
+  def change
+    create_table :rents do |t|
+      t.date :start_date, :null => false, :default => Time.now
```
reemember to use the new convention for the assignation, using `:` instead of `=>`  for example, `null: false`

leoncalermo commented on spec/factories/rent_factory.rb:
```diff
@@ -0,0 +1,10 @@
+FactoryBot.define do
+  factory :rent, class: Rent do
+    start_date { Time.zone.now }
+    end_date { Time.zone.now + (60 * 60 * 24) }
```
what ? 🤔  i dont undestand why the magic numbers 😆

leoncalermo commented on spec/models/rent_spec.rb:
```diff
@@ -0,0 +1,45 @@
+  it { is_expected.to validate_presence_of(:book) }
+  it { is_expected.to validate_presence_of(:user) }
+
+  describe '#save' do
```
same i commentes in the other PR, this is not necessary here, probably it would be better to do this kind of validations in the controller of this models when the time comes

leoncalermo commented on spec/models/rent_spec.rb:
```diff
@@ -0,0 +1,45 @@
+  end
+end
+
+describe Rent do
```
This is fine in this case  because you have specific validations that cant be validated with the common  validations

joredudiaz committed:
Changed magic numbers at rent_factory, remove unnecessary test at rent_spec, change hash style

joredudiaz committed:
Merge branch 'book_test' into rent

joredudiaz committed:
add default host to environment
